plugins/native-script-example/src/main/java/org/elasticsearch/examples/nativescript/script/CosineSimilarityScoreScript.java

```diff
@@ -0,0 +1,131 @@
+ * "Information Retrieval", Chapter 6, Eq. 6.12 (link:
+ * http://nlp.stanford.edu/IR-book/). This implementation only scores a list of
+ * terms on one field.
+ */
+public class CosineSimilarityScoreScript extends AbstractSearchScript {
+
+    // the field containing the terms that should be scored, must be initialized
+    // in constructor from parameters.
+    String field = null;
+    // terms that are used for scoring, must be unique
+    List<String> terms = null;
+    // weights, in case we want to put emphasis on a specific term. In the most
+    // simple case, 1.0 for every term.
+    List<Double> weights = null;
+
+    public static final String SCRIPT_NAME = "cosine_sim_script_score";
+
+    /**
+     * Factory that is registered in
+     * {@link org.elasticsearch.examples.nativescript.plugin.NativeScriptExamplesPlugin#onModule(org.elasticsearch.script.ScriptModule)}
+     * method when the plugin is loaded.
+     */
+    public static class Factory implements NativeScriptFactory {
+
+        /**
+         * This method is called for every search on every shard.
+         *
+         * @param params
+         *            list of script parameters passed with the query
+         * @return new native script
+         */
+        @Override
+        public ExecutableScript newScript(@Nullable Map<String, Object> params) {
+            return new CosineSimilarityScoreScript(params);
+        }
+
+        /**
+         * Indicates if document scores may be needed by the produced scripts.
+         *
+         * @return {@code true} if scores are needed.
+         */
+        @Override
+        public boolean needsScores() {
+            return false;
+        }
+
+        @Override
+        public String getName() {
+            return SCRIPT_NAME;
+        }
+    }
+
+    /**
+     * @param params
+     *            terms that a scored are placed in this parameter. Initialize
+     *            them here.
+     */
+    @SuppressWarnings("unchecked")
+    private CosineSimilarityScoreScript(Map<String, Object> params) {
+        params.entrySet();
+        // get the terms
+        terms = (List<String>) params.get("terms");
+        weights = (List<Double>) params.get("weights");
+        // get the field
+        field = (String) params.get("field");
+        if (field == null || terms == null || weights == null) {
+            throw new IllegalArgumentException("cannot initialize " + SCRIPT_NAME + ": field, terms or weights parameter missing!");
+        }
+        if (weights.size() != terms.size()) {
+            throw new IllegalArgumentException("cannot initialize " + SCRIPT_NAME + ": terms and weights array must have same length!");
+        }
+    }
+
+    @Override
+    public Object run() {
+        try {
+            float score = 0;
+            // first, get the ShardTerms object for the field.
+            IndexField indexField = this.indexLookup().get(field);
```

Comment:
It does. The example was never meant to "promote" scripts over custom similarities. IndexLookup makes it easy to test different methods before writing an actual plugin and this example shows how. However, we can add a "don't try this in production" warning.